src/bootstrap/src/core/build_steps/doc.rs

```diff
@@ -1195,6 +1195,11 @@ impl Step for UnstableBookGen {
     fn run(self, builder: &Builder<'_>) {
         let target = self.target;
 
+        let stage = builder.top_stage;
+        let compiler = builder.compiler(stage, builder.config.host_target);
```

Comment:
cc @Kobzol -- I imagine we'd prefer this to be prepare_doc_compiler and threaded in as a field of UnstableBookGen?